### PR TITLE
Emmett collisions

### PR DESCRIPTION
**Pull request by Emmett-Byrne**

### CODE REVIEW

File: ARGO/include/ColliderAABB.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+#include "Component.h"
+class ColliderAABB : public Component
```
Comment (CathalRedmond) [inline ARGO/include/ColliderAABB.h]: component should be added to end of name of class, all other components have it like this

File: ARGO/include/ColliderCircle.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+#include "Component.h"
+class ColliderCircle : public Component
```
Comment (CathalRedmond) [inline ARGO/include/ColliderCircle.h]: Component at end of class name

File: ARGO/include/ColliderCircle.h [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+{
+public:
+	ColliderCircle(int t_radius);
+	int getRadius();
+	void setRadius(int t_radius);
+private:
+	int m_radius;
```
Comment (CathalRedmond) [inline ARGO/include/ColliderCircle.h]: Any particular reason this isn't a float? AABB's bounds are floats

File: ARGO/include/CollisionSystem.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+#include "BaseSystem.h"
+#include "ColliderAABB.h"
+#include "ColliderCircle.h"
+#include "TransformComponent.h"
+#include "Utilities.h"
+#include "QuadTree.h"
+
+class CollisionSystem :	public BaseSystem
```
Comment (CathalRedmond) [inline ARGO/include/CollisionSystem.h]: big space between ':' and 'public'

File: ARGO/include/CollisionSystem.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+{
+public:
+	CollisionSystem();
+	void update(Entity& t_entity);
+	void handleCollisions();
+
+private:
+	bool circleToCircleCollision(Entity* t_entity1, Entity* t_entity2);
```
Comment (CathalRedmond) [inline ARGO/include/CollisionSystem.h]: Maybe change t_entity to circle to have the same as below function

File: ARGO/include/Game.h [hunk 6]
```diff
@@ -44,6 +47,7 @@ class Game
 	InputSystem m_inputSystem;
 	RenderSystem m_renderSystem;
 	AiSystem m_aiSystem;
+	CollisionSystem m_collsisionSystem;
```
Comment (CathalRedmond) [inline ARGO/include/Game.h]: misspelling - should be 'm_collisionSystem'

File: ARGO/src/CollisionSystem.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	{
+		TransformComponent* rectPosition = static_cast<TransformComponent*>(m_circleColliderBuffer[i]->getAllComps().at(COMPONENT_ID::TRANSFORM_ID));
+		int radius = static_cast<ColliderCircle*>(m_circleColliderBuffer[i]->getAllComps().at(COMPONENT_ID::COLLIDER_CIRCLE_ID))->getRadius();
+		glm::vec2 bounds{ radius, radius };
+		Quad quad{ NULL, rectPosition->getPos() - bounds, bounds * 2.0f };
+
+		std::vector<Entity*> entities;
+		m_quadTree.retrieve(&entities, quad);
+
+		for (auto& entity : entities)
```
Comment (CathalRedmond) [inline ARGO/src/CollisionSystem.cpp:46]: auto is ambigious

File: ARGO/src/CollisionSystem.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+		{
+			TransformComponent* otherPosition = static_cast<TransformComponent*>(entity->getAllComps().at(COMPONENT_ID::TRANSFORM_ID));
+			if (entity != m_circleColliderBuffer[i])
+			{
+				if (entity->getAllComps().at(COMPONENT_ID::COLLIDER_AABB_ID) && circleToAABBCollision(m_circleColliderBuffer.at(i), entity))
+				{
+					//handle aabb to circle collisisons here
```
Comment (CathalRedmond) [inline ARGO/src/CollisionSystem.cpp:53]: should this be here

File: ARGO/src/CollisionSystem.cpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+				}
+				else if (entity->getAllComps().at(COMPONENT_ID::COLLIDER_CIRCLE_ID) && circleToCircleCollision(m_circleColliderBuffer.at(i), entity))
+				{
+					int otherRadius = static_cast<ColliderCircle*>(entity->getAllComps().at(COMPONENT_ID::COLLIDER_CIRCLE_ID))->getRadius();
+					count++;
+					glm::vec2 between = rectPosition->getPos() - otherPosition->getPos();
```
Comment (CathalRedmond) [inline ARGO/src/CollisionSystem.cpp]: between what?

File: ARGO/src/CollisionSystem.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+						between = glm::vec2(m_seperationScaler, m_seperationScaler);
+					}
+					rectPosition->addPos(between);
+					otherPosition->addPos(-between);
+				}
+			}
+		}
+	}
+	m_circleColliderBuffer.clear();
+	std::cout << m_quadTree.countAll(0) << std::endl;
```
Comment (CathalRedmond) [inline ARGO/src/CollisionSystem.cpp]: if this was for testing it should be gotten rid of

Comment (CathalRedmond) [reply ARGO/src/CollisionSystem.cpp]: or inside #ifdef debug.

File: ARGO/src/CollisionSystem.cpp [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+bool CollisionSystem::circleToCircleCollision(Entity* t_entity1, Entity* t_entity2)
+{
+	glm::vec2 position1 = static_cast<TransformComponent*>(t_entity1->getAllComps().at(COMPONENT_ID::TRANSFORM_ID))->getPos();
+	int radius1 = static_cast<ColliderCircle*>(t_entity1->getAllComps().at(COMPONENT_ID::COLLIDER_CIRCLE_ID))->getRadius();
+	glm::vec2 position2 = static_cast<TransformComponent*>(t_entity2->getAllComps().at(COMPONENT_ID::TRANSFORM_ID))->getPos();
+	int radius2 = static_cast<ColliderCircle*>(t_entity2->getAllComps().at(COMPONENT_ID::COLLIDER_CIRCLE_ID))->getRadius();
+
+	glm::vec2 between = position1 - position2;
```
Comment (CathalRedmond) [inline ARGO/src/CollisionSystem.cpp]: again, between what?

File: ARGO/src/CollisionSystem.cpp [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+bool CollisionSystem::circleToAABBCollision(Entity* t_entityCircle, Entity* t_entityAABB)
+{
+	glm::vec2 circlePosition = static_cast<TransformComponent*>(t_entityCircle->getAllComps().at(COMPONENT_ID::TRANSFORM_ID))->getPos();
+	int radius = static_cast<ColliderCircle*>(t_entityCircle->getAllComps().at(COMPONENT_ID::COLLIDER_CIRCLE_ID))->getRadius();
+	glm::vec2 rectPosition = static_cast<TransformComponent*>(t_entityAABB->getAllComps().at(COMPONENT_ID::TRANSFORM_ID))->getPos();
+	glm::vec2 bounds = static_cast<ColliderAABB*>(t_entityAABB->getAllComps().at(COMPONENT_ID::COLLIDER_AABB_ID))->getBounds();
+
+	int DeltaX = circlePosition.x - std::max(rectPosition.x, std::min(circlePosition.x, rectPosition.x + bounds.x));
```
Comment (CathalRedmond) [inline ARGO/src/CollisionSystem.cpp]: small d in deltaX and deltaY

File: ARGO/src/Game.cpp [hunk 13]
```diff
@@ -53,7 +51,8 @@ Game::Game() :
 		{
 			m_entities.emplace_back();
 			m_entities.at(i).addComponent(new TransformComponent());
-			m_entities.at(i).addComponent(new AiComponent());
+			//m_entities.at(i).addComponent(new AiComponent());
```
Comment (CathalRedmond) [inline ARGO/src/Game.cpp]: should this be here?

File: ARGO/src/PhysicsSystem.cpp [hunk 14]
```diff
@@ -35,17 +35,17 @@ void PhysicsSystem::checkBorder(TransformComponent* t_pos)
 {
 	//magic numbers for window size will be replaced for world size
 	//once we have a class that holds globals
```
Comment (CathalRedmond) [inline ARGO/src/PhysicsSystem.cpp]: this comment can be removed

File: ARGO/src/QuadTree.cpp [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+QuadTree::~QuadTree()
+{
+	m_objects.clear();
+	m_nodes.clear();
+}
+
+void QuadTree::clear()
+{
+	m_objects.clear();
```
Comment (CathalRedmond) [inline ARGO/src/QuadTree.cpp:18]: this seems to do the same as ~QuadTree() why not call this inside that function

File: ARGO/src/QuadTree.cpp [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+				iter++;
+			}
+		}
+	}
+}
+
+void QuadTree::retrieve(std::vector<Entity*>* t_vector, Quad t_data)
+{
+	int index = getIndex(t_data);
+	if (index != -1 && !m_nodes.empty()) {
```
Comment (CathalRedmond) [inline ARGO/src/QuadTree.cpp]: '{' bracket should be moved down one line

File: ARGO/src/QuadTree.cpp [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+	float verticalMidpoint = m_boundsTopLeft.x + (m_boundsBottomRight.x / 2);
+	float horizontalMidpoint = m_boundsTopLeft.y + (m_boundsBottomRight.y / 2);
+
+	// Object can completely fit within the top quadrants
+	bool topQuadrant = (t_data.position.y + t_data.size.y <= horizontalMidpoint);
+	// Object can completely fit within the bottom quadrants
+	bool bottomQuadrant = (t_data.position.y >= horizontalMidpoint);
+
+	// Object can completely fit within the left quadrants
+	if (t_data.position.x + t_data.size.x <= verticalMidpoint) {
```
Comment (CathalRedmond) [inline ARGO/src/QuadTree.cpp]: '{' should be moved down

Comment (CathalRedmond) [reply ARGO/src/QuadTree.cpp]: same on line 98 and 101

File: ARGO/src/QuadTree.cpp [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+		if (topQuadrant) {
+			index = 1;
+		}
+		else if (bottomQuadrant) {
+			index = 2;
+		}
+	}
+	// Object can completely fit within the right quadrants
+	else if (t_data.position.x >= verticalMidpoint) {
```
Comment (CathalRedmond) [inline ARGO/src/QuadTree.cpp]: bracket {

Comment (CathalRedmond) [reply ARGO/src/QuadTree.cpp]: again on line 107 and 110

File: ARGO/src/QuadTree.cpp [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+			index = 3;
+		}
+	}
+
+	return index;
+}
+
+void QuadTree::split()
+{
+	int subWidth = (m_boundsBottomRight.x / 2);
```
Comment (CathalRedmond) [inline ARGO/src/QuadTree.cpp]: what does sub mean?

**Review by CathalRedmond — CHANGES_REQUESTED** [review]

File: ARGO/src/CollisionSystem.cpp (hunk 8 above)
Comment (Violet-Byrne) [reply ARGO/src/CollisionSystem.cpp:53]: yes

**Review by CathalRedmond — APPROVED** [review]

**Pull request merged**